eknrenderer/eknr-renderer.c

```diff
@@ -0,0 +1,1306 @@
+/* Copyright 2018 Endless Mobile, Inc. */
+
+#include "config.h"
+
+#include <string.h>
+#include <stdlib.h>
+
+#include <endless/endless.h>
+#include <glib/gi18n.h>
+#include <json-glib/json-glib.h>
+#include <mustache.h>
+
+#include "eknr-boxed-set-info.h"
+#include "eknr-errors.h"
+#include "eknr-renderer.h"
+
+/**
+ * SECTION:renderer
+ * @title: Renderer
+ * @short_description: A post-processing renderer for HTML article content
+ *
+ * The renderer is responsible for adding some final postprocessing to
+ * articles on the client side before it is displayed to the user. What
+ * postprocessing happens depends on the article's source.
+ */
+struct _EknrRenderer
+{
+  GObject parent_instance;
+};
+
+typedef struct _EknrRendererPrivate
+{
+  GHashTable *cache;
+} EknrRendererPrivate;
+
+G_DEFINE_TYPE_WITH_PRIVATE (EknrRenderer,
+                            eknr_renderer,
+                            G_TYPE_OBJECT)
+
+typedef struct _RendererMustacheData {
+  GVariantDict  *variables;
+  GError       **error; /* non-owned */
```

Comment:
It's unusual enough to keep this in a struct that I'd add some clarification in a comment. After staring at the code for a bit I understand that it's because `mustache_render` gives you one pointer for all the things that need to be passed to your vfuncs, including an error-out location. Maybe a comment above the struct explaining that that's what it's for?